db/transforms/base.py

```diff
@@ -10,13 +11,39 @@
 from db.records.operations import group, relevance, sort as rec_sort
 
 
+class UniqueConstraintMapping(NamedTuple):
```

Comment:
The main reason I used the "typed" variant `NamedTuple` is to have this data structure look like any other in our code base, which also includes having a nice, consistent place to put the docstring. I think the consistency is nice when we can have it.

I don't use the `__doc__` attribute. But, my IDE knowing about the attribute names (`input_alias` and `output_alias`) is nice though; don't get that when using `namedtuple`. `NamedTuple` in general is friendlier to IDEs.

I put the type annotations there for no particular reason. The interepreter seems to force me to do it. Though, I like that I could use the `Optional` annotation: documents at a glance that the attribute can be `None`.

If you care about this, could you explain why you prefer `namedtuple` over `NamedTuple`?

Reply:
1. `namedtuple` is the "normal" syntax. It's used in docs and the vast majority of references to 'named tuples in python'. For evidence, see:
   - other places we use named tuples in our code (all use the generator, as is standard)
   - other places on github using named tuples in python
   - Google search results for namedtuple (or NamedTuple)
   - a cursory search of the python docs for named tuples and their usage.

   This is important, since it's an open source project (and also, why use non-standard syntax "to make it look nice"? That's a little upside-down to me).
2. The class syntax you've chosen introduces type hints for no good reason
3. `NamedTuple`s are _tuples_ first. They have an index. They're immutable. The class syntax obfuscates this. In particular, the index of different attributes is set by the line order in which they're defined in the class, which is horrible.
4. Things defined via `class` syntax in python usually come with certain things (e.g., a `__dict__`) that `NamedTuple`s lack, since they're actually code generators.
5. Anyone who cares about performance will have to go deep figuring out the differences between what's generated by the `class` syntax vs. the factory function (if any). The `class` syntax tricks you into thinking this is a "standard python object" with all the baggage that implies (it's not, but the syntax obfuscates that).

While I also wish there was a tidier way to get the docstring added using the standard syntax, I consider that subordinate to all the other concerns.